MrCyjaneK/monero_c
# Rust bindings

Proof of concept Rust bindings.  TODO:
- [x] Prove actual wallet usage (gen. seed or similar)
- [x] Rust wrapper structure(s) around the C FFI (*eg.* a struct to store a wallet with the C pointer)
- [x] Convert example from a bin to a lib
- [x] Add docs tests
- [x] Add integration tests
- [x] Improve error handling: add `get_status`
- [ ] Finish wrapping bindings
  - [ ] At minimum wrap all functions listed by binarybaron [here](https://github.com/binarybaron):
    - [x] get_address(&self, account_index: u32) -> GetAddress;
    - [x] get_balance(&self, account_index: u32) -> GetBalance;
    - [x] create_account(&self, label: String) -> CreateAccount;
    - [x] get_accounts(&self, tag: String) -> GetAccounts;
    - [x] open_wallet(&self, filename: String) -> WalletOpened;
    - [x] close_wallet(&self) -> WalletClosed;
    - [x] create_wallet(&self, filename: String, language: String) -> WalletCreated;
    - [x] transfer(&self, account_index: u32, destinations: Vec<Destination>, get_tx_key: bool) -> Transfer;
    - [x] get_height(&self) -> BlockHeight;
    - [x] check_tx_key(&self, txid: String, tx_key: String, address: String) -> CheckTxKey;
    - [x] generate_from_keys(&self, filename: String, address: String, spendkey: String, viewkey: String, restore_height: u32, password: String, autosave_current: bool) -> GenerateFromKeys;
    - [x] refresh(&self) -> Refreshed;
    - [x] sweep_all(&self, address: String) -> SweepAll;
    - [ ] check_tx_key(&self) -> Version;
- [ ] Test wrappers
  - [ ] transfer
  - [ ] sweep_all
  - [ ] check_tx_key
- [x] Fix example for subsequent runs (fails when wallet file already exists)
- [ ] Address `clippy` warnings (format, etc.)

Ensure cross-platform compatibility:

and so we don't overload our upstream:
- [ ] Rework PR as the addition of a git subtree of a standalone repo with a build script (preferably the upstream's)

We may also want to: (optional)
- Allow the wallet files storage path to be configured
- Add coin control
- Add hardware wallets

~~Optionally, improve the build script:~~
~~- [x] Don't clone a brand new repo, build in place (unless passed `-c|--clean`)~~
*Build script removed*
